versioned_hdf5/subchunk_map.py

```diff
@@ -180,145 +576,87 @@ def as_subchunk_map(
-            )
-            mask = mask.reshape((mask.shape[0] // n, n))
-
-            # count how many elements were selected in each chunk
-            chunk_selected_counts = np.sum(mask, axis=1, dtype=np.intp)
-
-            # compute offsets based on selected counts which will be used to build
-            # the masks for each chunk
-            chunk_selected_offsets = np.zeros(
-                len(chunk_selected_counts) + 1, dtype=np.intp
-            )
-            chunk_selected_offsets[1:] = np.cumsum(chunk_selected_counts)
-
-            # chunk_idxs for the chunks which are not empty
-            chunk_idxs = np.flatnonzero(chunk_selected_counts)
-
-            for chunk_idx in chunk_idxs:
-                chunk_start = chunk_idx * n
-                chunk_stop = min((chunk_idx + 1) * n, d)
-                chunk_subindexes_for_axis.append(
-                    (
-                        Slice(chunk_start, chunk_stop, 1),
-                        np.concatenate(
-                            [
-                                np.zeros(chunk_selected_offsets[chunk_idx], dtype=bool),
-                                np.ones(chunk_selected_counts[chunk_idx], dtype=bool),
-                                np.zeros(
-                                    chunk_selected_offsets[-1]
-                                    - chunk_selected_offsets[chunk_idx + 1],
-                                    dtype=bool,
-                                ),
-                            ]
-                        ),
-                        np.flatnonzero(i.array[chunk_start:chunk_stop]),
-                    )
-                )
+            mapper = BooleanArrayMapper(i.raw, n, d)
         elif isinstance(i, Integer):
-            i_raw: Py_ssize_t = i.raw
-            chunk_idx = i_raw // n
-            chunk_start = chunk_idx * n
-            chunk_stop = min((chunk_idx + 1) * n, d)
-            chunk_subindexes_for_axis.append(
-                (
-                    Slice(chunk_start, chunk_stop, 1),
-                    (),
-                    i_raw - chunk_start,
-                )
-            )
+            mapper = IntegerMapper(i.raw, n, d)
         else:
             raise NotImplementedError(f"index type {type(i)} not supported")
-
-        chunk_subindexes.append(chunk_subindexes_for_axis)
+        mappers.append(mapper)
 
     # Handle the remaining suffix axes on which we did not select, we still need to
     # break them up into chunks.
     for n, d in zip(suffix_chunk_size, suffix_shape):
-        chunk_slices_gen = (
-            Slice(chunk_idx * n, min((chunk_idx + 1) * n, d), 1)
-            for chunk_idx in range((d + n - 1) // n)
-        )
-        chunk_subindexes.append(
-            [(chunk_slice, chunk_slice.raw, ()) for chunk_slice in chunk_slices_gen]
-        )
+        mapper = EverythingMapper(n, d)
+        mappers.append(mapper)
+
+    return idx, mappers
+
+
+def as_subchunk_map(
```

Comment:
This is just a temporary refactor, for my sanity to verify that I didn't break anything fundamental early on, and for the sanity of reviewers. It will be completely removed in the final version.